Classes/Middleware/FrontendUserAuthenticator.php

```diff
@@ -0,0 +1,158 @@
+    {
+        if (Util::getIsTYPO3VersionBelow10()
+            || !$request->hasHeader(PageIndexerRequest::SOLR_INDEX_HEADER)
+        ) {
+            return $handler->handle($request);
+        }
+
+        // disable TSFE cache for TYPO3 v10
+        $request = $request->withAttribute('noCache', true);
+        $jsonEncodedParameters = $request->getHeader(PageIndexerRequest::SOLR_INDEX_HEADER)[0];
+
+        /* @var PageIndexerRequestHandler $pageIndexerRequestHandler */
+        $pageIndexerRequestHandler = GeneralUtility::makeInstance(PageIndexerRequestHandler::class, $jsonEncodedParameters);
+        if (!$pageIndexerRequestHandler->getRequest()->isAuthenticated()) {
+            /* @var SolrLogManager $logger */
+            $logger = GeneralUtility::makeInstance(SolrLogManager::class, self::class);
+            $logger->log(
+                SolrLogManager::ERROR,
+                'Invalid Index Queue Frontend Request detected!',
+                [
+                    'page indexer request' => (array)$pageIndexerRequestHandler->getRequest(),
+                    'index queue header' => $jsonEncodedParameters
+                ]
+            );
+
+            return new JsonResponse(['error' => ['code' => 403, 'message' => 'Invalid Index Queue Request.']], 403);
+
+        }
+        $request = $this->tryToAuthenticateFrontendUser($pageIndexerRequestHandler, $request);
+
+        return $handler->handle($request);
+    }
+
+    /**
+     * Fakes a logged in user to retrieve access restricted content.
+     *
+     * @param PageIndexerRequestHandler $handler
+     * @param ServerRequestInterface $request
+     * @return ServerRequestInterface
+     * @noinspection PhpUnused
+     */
+    protected function tryToAuthenticateFrontendUser(PageIndexerRequestHandler $handler, ServerRequestInterface $request): ServerRequestInterface
+    {
+        $accessRootline = $this->getAccessRootline($handler);
+        $stringAccessRootline = (string)$accessRootline;
+
+        if (empty($stringAccessRootline)) {
+            return $request;
+        }
+
+        $groups = $accessRootline->getGroups();
+
+        /* @var FrontendUserAuthentication $feUser */
+        $feUser = GeneralUtility::makeInstance(FrontendUserAuthentication::class);
+        $feUser->user[$feUser->username_column] = AuthorizationService::SOLR_INDEXER_USERNAME;
+        /* @noinspection PhpParamsInspection */
+        $this->context->setAspect('frontend.user', GeneralUtility::makeInstance(UserAspect::class, $feUser, $groups));
+        $request = $request->withAttribute('frontend.user', $feUser);
+
+        return $request;
+    }
+
+    /**
+     * Gets the access rootline as defined by the request.
+     *
+     * @param PageIndexerRequestHandler $handler
+     * @return Rootline The access rootline to use for indexing.
+     */
+    protected function getAccessRootline(PageIndexerRequestHandler $handler): ?Rootline
```

Comment:
```suggestion
    protected function getAccessRootline(PageIndexerRequestHandler $handler): Rootline
```

shouldn't that always return a Rootline object